Add time needed setting with filter

boblinthorst opened the pull request:
## Summary

This PR can be summarized in the following changelog entry:

* Adds a setting and filter allowing users to edit the text describing the needed time for a how-to guide.

## Relevant technical choices:

* Added a filter to allow to change the text via filter.
* Used the same structure as existing settings for the how-to block.
* Prioritised the user-input in the settings over the filter.

## Test instructions

This PR can be tested by following these steps:

* Checkout this branch.
* build this branch `yarn, grunt build:js && grunt build:css && yarn start`
* Create a new post. add a how-to block to the new post. 
* Check that the text before the needed time reads `time needed:`.
* go to the settings of the how-to block.
* Check that the default text of `time needed:` is in the input field as a placeholder.
* Change the placeholder text in the settings.
* Check that the text is changed to the user input in the how-to block.
* Save draft, Check that the text is changed to the user input on the post preview.

To Test the filter:
* Go to `js/src/structured-data-blocks/how-to/block.js` and add the following:
```
import get from "lodash/get";
```
and inside the `edit` function before tue return of `<howTo`:
```
const addFilter = get( window, "wp.hooks.addFilter" );
addFilter( "wpseo_duration_text", "random", ( defaultValue ) => {
	return "ReplacedViaFilter";
} );
```
* Go to a (new) post-edit page.
* Add a how-to block ( delete the one if it already exists in the post), one with the default value is needed.
* Add a duration to the how-to.
* Check that the text for the duration time is `ReplacedViaFilter` (changed by the filter).
* Change the text in the settings of the how-to page.
* Check that the text for the duration time is changed in the block.
* Save draft, check that the text for the duration time is changed in the preview of the post.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes #10665

afercia commented on js/src/structured-data-blocks/how-to/components/HowTo.js:
```diff
@@ -40,6 +42,8 @@ export default class HowTo extends Component {
 	constructor( props ) {
 		super( props );
 
+		const applyFilters = get( window, "wp.hooks.applyFilters", noop );
```
question: I'm not familiar with the way lodash/get works, but is this basically equivalent to getting something from the global `window.wp` without checking if it exists before?

boblinthorst replied:
What happens is that, if `wp.hooks.aplyFilters` doesn't exist, `noop` is used as a fallback. Right now, if it isn't found, the defaultDurationText becomes empty as a result because the default of applyFilters isn't set as well. so I need to address that.

afercia commented on js/src/structured-data-blocks/how-to/components/HowTo.js:
```diff
@@ -40,6 +42,8 @@ export default class HowTo extends Component {
 		super( props );
 
+		const applyFilters = get( window, "wp.hooks.applyFilters", noop );
+		const defaultDurationText = applyFilters( "wpseo_duration_text", __( "Time needed:" ) );
```
missing domain "wordpress-seo"?

afercia commented on js/src/structured-data-blocks/how-to/components/HowTo.js:
```diff
@@ -50,6 +54,11 @@ export default class HowTo extends Component {
 		this.toggleListType  = this.toggleListType.bind( this );
+		this.setDurationText = this.setDurationText.bind( this );
+
+		if( ! this.props.attributes.durationText ) {
```
missing space after if

afercia commented on js/src/structured-data-blocks/how-to/components/HowTo.js:
```diff
@@ -443,7 +455,7 @@ export default class HowTo extends Component {
 					<p className="schema-how-to-total-time">
-						{ __( "Time needed:", "wordpress-seo" ) }
+						{ durationText }
 						&nbsp;
```
since we now allow the timeString to wrap in a new line, I think this `&nbsp;` should be removed

afercia commented on js/src/structured-data-blocks/how-to/components/HowTo.js:
```diff
@@ -443,7 +455,7 @@ export default class HowTo extends Component {
-						{ __( "Time needed:", "wordpress-seo" ) }
+						{ durationText }
 						&nbsp;
 						{ timeString }.
```
Instead, we should consider to wrap timeString in a span element with a CSS class, to allow theme developers to style it (see screenshot I'll post in the extended comment)

afercia commented:
CR 🚧 

Apart from the points mentioned above, I've noticed a few issues:

1
It is now possible to completely remove the "Time needed:" text just emptying the field in the sidebar:

<img width="330" alt="screen shot 2018-09-05 at 12 58 53" src="https://user-images.githubusercontent.com/1682452/45091826-e50b0600-b113-11e8-9b24-d784de72e19f.png">

In the editor, this text is used as a fieldset legend and should not be removed. Also, when saving and publishing there's no text in the front-end too. I guess this is something we want to avoid. Maybe consider to always restore the default text.

2
as noted in the comments, the `&nbsp;` should be removed and I'd consider to wrap the `timeString` text in a span element with a CSS class to allow theme developers so style it if they want. Useful when the custom text is very long. In the example below, a CSS class would allow to have a better styling:

<img width="584" alt="screen shot 2018-09-05 at 12 53 13" src="https://user-images.githubusercontent.com/1682452/45092030-7ed2b300-b114-11e8-836b-3c151929f5d7.png">

3
I've noticed something weird when using the filter, to reproduce:
- set the filter as described in this PR, then build the JS
- create a new post, add a How-to block and set the time needed
- save the post
- remove the How-to block
- add a new one
- click the button to set the Time needed
- at this point, I see a very long string with the `newAddition` text added by the filter multiple times:

<img width="886" alt="screen shot 2018-09-05 at 12 59 59" src="https://user-images.githubusercontent.com/1682452/45092153-dcff9600-b114-11e8-9cf2-b7cea4445348.png">

boblinthorst commented:
Concerning the wierd bug. This is probably due to my quick example not being the greatest. I added the addFilter in a place where you probably shoudn't (but it was only for quick testing purposes, so i didn't mind too much). Because of it, it seems multiple filters were created, all resulting in adding a tiny bit of content to the string. 

My add_filter testing code example was simply not what it should be, I've replaced it.

boblinthorst commented:
I added the span with a css class. Styling it to add space between the text and the actual numbers gave me some issues though that I can't quite explain. Suggestions are welcome here. For now, I put the `&nbsp` back

andizer commented:
CR done

Dieterrr commented:
Acceptance done 👍
